### DISCUSSION TITLE
Creating Thin volumes using the lvcreate-options

### DISCUSSION THREAD

**Discussion by sp98** [topolvm/topolvm#439] [Q&A] [answered]
I'm trying to create thin logical volumes using the new `lvcreate-options` but I don't think the new `lvcreate-options` is the best way to create thin volumes due to following reasons:

1. Thin-lv creation is two step process. First is to create the thin-pool and then to create actual LV. Although there is single command to create both thin-pool and thin-lv (`lvcreate -L 100M -T vg001/mythinpool -V1G -n thinvolume`), we need to always ensure that thin-pool is not created already, else the LV creation will fail. This would involve parsing the `lvcreate-options` to fetch the thin-pool name and checking if the thin-pool is already created or not. This is not a good option.

2. While creating a thin pool, we need to provide actual size of the thin-pool as well. It might be very convenient for the user to provide the size a string args in `lvcreate-options`. 

One solution for both the problems is to **parse the `lvcreate-options` string array and get the thin-pool details. But this won't be a very good solution.** 

Another option would be add a new API field in the `DeviceClass`. Something like:
 ```
 device-classes: 
  - default: true
    thin-pool:
    	name: mythinpool
    	size: <Some % of the available space on the VolumeGroup>
    name: ssd
    spare-gb: 10
``` 
Let me know what you think about this. @lentzi90 @satoru-takeuchi @toshipp

**Comment by toshipp** [comment]
I hand over this discussion to reviewers of #433 (@cupnes and @satoru-takeuchi). Please wait for a while.

**Comment by cupnes** [comment]
@sp98 
We consider it's better to add new fields to `DeviceClass` to support thin volumes. As the first step, could you write down the proposal of this feature like [this?](https://github.com/topolvm/topolvm/blob/main/docs/proposals/lvcreate-options.md)

Here are my random ideas about the design.

- The admin is in charge of creating thin pools beforehand.
- Get the capacities of each thin pool periodically and export these numbers as annotations of Node resources.

@lentzi90 
We'd like to support the thin snapshots feature in the future. Do you have any comments?

**Comment by lentzi90** [comment]
I think what @cupnes said makes sense.
Unfortunately I don't think I'll be able to work on it at this time. My employer hasn't expressed interest in thin volumes or snapshots. But I'll let you know if things change! :slightly_smiling_face:

**Reply by sp98** [reply]
We have requirements for thin volumes and snapshots.  We will be happy to work on the design and development efforts.

**Comment by toshipp** [comment]
It is a just idea and more work needs but it may be more suitable. Why don't you introduce a more generic option like `lvbackend`?
The current `device-class` option heavily ties with VG, introducing thinp backend may introduce confusion into the codebase.
The new option will be something like:
```
lvbackend:
  - name: thinp
     thin-pool:
       pool-name: foo
  - name: vg
     vg:
        vg-name: bar
```

A possible implementation plan is:
1. implement thinp backend on lvbackend option and release the new version.
2. implement VG backend on lvbackend option and deprecate device-classes.
3. maybe after v1.0 release, drop device-classes entirely.

**Reply by toshipp** [reply]
Sorry, I missed that creating a thinp volume requires a VG name, so adding a new field is sufficient.

**Reply by leelavg** [reply]
- Is there any [limitation](https://github.com/topolvm/topolvm/blob/main/docs/proposals/lvcreate-options.md#should-multiple-device-classes-be-able-to-target-the-same-vg) that a device-class to volume group can only be one-to-one mapping
- We are writing the proposal doc as of now and wanted to know if there's any limitation as such
- To make things easier and safer we are thinking to have different device-class targeting a single volume group, one marked for thick lvs and another for thinpool which also eases up storage class mapping

@toshipp ^^

ps: i'm with @sp98

**Reply by cupnes** [reply]
@leelavg 
Yes, a device-class to volume group is one-to-one mapping. This limitation is still valid.

**Reply by leelavg** [reply]
hi @cupnes, would you please let me know or point me why device-class to volume group should be one-to-one mapping.

**Reply by cupnes** [reply]
@leelavg 
We were thinking of allowing the volume group name in the StorageClass, but that would show the internal implementation to the Kubernetes cluster, so we put the device-class in between. Changing that would have been a big change, so we restricted it to one-to-one correspondence.
On the other hand, StorageClass and device-class can be mapped N to 1. It means that a device-class can be used from some storage classes.

**Comment by toshipp** [comment, marked as answer]
Thin LV was implemented by https://github.com/topolvm/topolvm/pull/448
